lib/AST/TypeCheckRequests.cpp

```diff
@@ -546,3 +546,27 @@ bool DefaultTypeRequest::getPerformLocalLookup(const KnownProtocolKind knownProt
     default: return false;
   }
 }
+
+//----------------------------------------------------------------------------//
+// StructuralTypeRequest.
+//----------------------------------------------------------------------------//
+
+void StructuralTypeRequest::diagnoseCycle(DiagnosticEngine &diags) const {
+  diags.diagnose(SourceLoc(), diag::circular_reference);
+}
+
+void StructuralTypeRequest::noteCycleStep(DiagnosticEngine &diags) const {
+  diags.diagnose(SourceLoc(), diag::circular_reference_through);
+}
+
+Optional<Type> StructuralTypeRequest::getCachedResult() const {
+  auto decl = std::get<0>(getStorage());
+  if (decl->getUnderlyingTypeLoc().getType())
```

Comment:
We don't want to cache it here because this is where the interface type will be stored. Instead, you can change CacheKind::SeparatelyCached to CacheKind::Cached so that the request-evaluator uses its own caching.